Bugfix: allow missing modules if they are blacklisted

scheibelp opened the pull request:
Fixes #13346.

For `spack module loads` and `spack module find`, allow missing upstream modules (print a debug message vs. raising an error); also skip local modules if they are blacklisted.

Old behavior:

* `spack module find` will report an error if the root spec's module or any dependency module does not exist (regardless of whether it is blacklisted)
* `spack module loads` works the same

New behavior: the short version is that the module finding is now more lenient (especially for blacklisted modules):

* `spack module find` will report an error if the root spec's module does not exist, unless
  * the module for the spec is blacklisted
    * For the root package, a debug message is output so that the user could check this with `spack -d...`
    * For any dependency package (when requesting dependency modules with `spack module find`) a message will not be reported (it's an error if the module is not blacklisted, but if it is blacklisted, the module is silently skipped)
  * the spec is associated with an upstream
    * instead of throwing exceptions when the module isn't find, the logic now prints out a debug-level message so the user has a hint of what is missing if they were expecting to see a module file for an upstream package
* `spack module loads` works exactly the same way as the new `spack module find`, but informational messages are not generated for blacklisted packages

scheibelp commented:
@tgamblin this should add tests (e.g. for `spack module find` when there is a blacklist) but do you agree with the intended behavior in the PR description?

tgamblin commented:
@glennpj: does this fix #13346 for you?

glennpj commented:
> @glennpj: does this fix #13346 for you?

Yes, it does.

alalazo commented on lib/spack/spack/modules/common.py:
```diff
@@ -312,31 +312,51 @@ def upstream_module(self, spec, module_type):
+    if it is missing it is assumed to have been omitted purposefully. If the
+    module is not available (and it is not required or it is blacklisted), then
+    this function returns ``None``.
+    """
```
Minor OCD. Can we reformat to respect the [Google Style Guide](http://google.github.io/styleguide/pyguide.html#38-comments-and-docstrings), as we do in many other places?

tgamblin replied:
Is there a checker for this?

alalazo replied:
Not sure, really. There's a checker for [PEP257](http://www.pydocstyle.org/en/4.0.1/release_notes.html), but they seems to have only "initial support" for Google's conventions.

scheibelp commented:
@tgamblin changes are now addressed:

* explicit return of `None`
* updated docstring (per https://github.com/spack/spack/pull/13540#discussion_r353805350)
* `loads` output includes comments for missing (blacklisted/upstream) modules (note that the comment mentions that it could be blacklisted or that it could be missing from an upstream but doesn't distinguish) as well as an overall warning message
